### ISSUE TITLE
Compiler Hacks break Advanced Optimization if you refresh namespaces (relatively easy fix identified)

### ISSUE THREAD

**Issue by awkay** [fulcrologic/fulcro#67]
The workaround is to make sure tools-namespace is around when you're doing advanced compilation.

The fix is to figure out how to keep Closure from removing the `static` methods on the generated React classes. 

Some possibilities:
1. Figure out an alternate way of implementing `static`...perhaps it should just be metadata in cljs and clj, but that means some internal API changes that could break existing programs (anything that has been trying to call those statically against the classes)
2. Figure out why it breaks with respect to Closure and submit a patch there?

**Comment by awkay** [comment]
So, I've placed a TODO item in defui. I think if we statically call the static methods just after we defined them, then Closure won't remove them. The reason they are getting removed is it cannot tell they are getting used because they are always called through variables (instead of the hard class name itself). If we throw that in a try/catch, it should make it so we don't need the compiler hack anymore!

**Comment by awkay** [comment]
@currentoor If you're still looking for something to do, this one is really tractable.

**Comment by currentoor** [comment]
@awkay thanks, I'll take a look after I finish working on porting fulcro-inspect to chrome devtools.

**Comment by awkay** [comment]
Cool!

**Comment by awkay** [comment]
So, I think this is fixed on branch feature/issue-67-compiler-hack. I think it deserves some serious testing, but it works for me on the dev guide adv compile build.

**Comment by wilkerlucio** [comment]
This seems to work, but I got some weird warnings:

```
Dec 21, 2017 9:01:46 PM com.google.javascript.jscomp.LoggerErrorManager println
WARNING: /Users/wilkerlucio/Development/shuffle-modules/shuffle.common/resources/public/js/compiled/advanced/nubank/shuffle/common/fulcro/portal.js:596: WARNING - unreachable code
try{}catch (e61858){var e_61903 = e61858;
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Dec 21, 2017 9:01:46 PM com.google.javascript.jscomp.LoggerErrorManager println
WARNING: /Users/wilkerlucio/Development/shuffle-modules/shuffle.common/resources/public/js/compiled/advanced/nubank/shuffle/common/ui/clipboard.js:253: WARNING - unreachable code
try{}catch (e58645){var e_58738 = e58645;
     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by awkay** [comment]
OH, I didn't account for the case where a component had no protocols on it.

**Comment by awkay** [comment]
Fixed and merged.

**Comment by awkay** [comment]
Looks like I missed something. I forgot to remove the hacks, and the try/catch isn't being emitted.

**Comment by awkay** [comment]
Fixed in 2.1.2
